models/issue_stress.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// TestIssueNoDupIndex Performs a stress test of the INSERT ... SELECT function of database for inserting issues
+func TestIssueNoDupIndex(t *testing.T) {
+	assert.NoError(t, PrepareTestDatabase())
+
+	const initialIssueFill = 1000 // issues inserted prior to stress test
+	const maxTestDuration = 60    // seconds
+	const threadCount = 8         // max simultaneous threads
+
+	var err error
+	repo := AssertExistsAndLoadBean(t, &Repository{ID: 3}).(*Repository)
+	doer := AssertExistsAndLoadBean(t, &User{ID: repo.OwnerID}).(*User)
+
+	// Pre-load
+	for i := 1; i < initialIssueFill; i++ {
+		issue := &Issue{
+			RepoID:   repo.ID,
+			PosterID: repo.OwnerID,
+			Index:    int64(i + 5000), // Avoid clashing with other tests
+			Title:    fmt.Sprintf("NoDup initial %d", i),
+		}
+		_, err = x.Insert(issue)
+		assert.NoError(t, err)
+	}
+
+	fmt.Printf("TestIssueNoDupIndex(): %d rows created\n", initialIssueFill)
+
+	until := time.Now().Add(time.Second * maxTestDuration)
+
+	var hasErrors int32
+	var wg sync.WaitGroup
+
+	f := func(thread int) {
+		defer wg.Done()
+		sess := x.NewSession()
+		defer sess.Close()
+		i := 1
+		for {
+			if time.Now().After(until) || atomic.LoadInt32(&hasErrors) != 0 {
+				return
+			}
+			issue := &Issue{
```

Comment:
Why not add transaction here?

Reply:
@lunny You mean database transactions? I think it will cause a serious escalation problem (i.e. table wide locks).

Reply:
Sorry, now I get what you mean. I'll test it right now.

Reply:
Unfortunately, adding transactions did not make any difference. The CI tests seem to be down ATM, but I've tested with MSSQL in my server and I've got the same results.